src/collisions.rs

```diff
@@ -0,0 +1,462 @@
+            && other.top() < self.bottom()
+            && self.top() < other.bottom()
+    }
+
+    pub fn union(&self, other: &Rect) -> Rect {
+        let x = self.x.min(other.x);
+        let y = self.y.min(other.y);
+
+        Rect::new(
+            x,
+            y,
+            self.right().max(other.right()) - x,
+            self.bottom().max(other.bottom()) - y,
+        )
+    }
+
+    pub fn union_vec2(&self, vec: &Vec2) -> Rect {
+        let rect = Rect::new(vec.x, vec.y, 0.0, 0.0);
+        self.union(&rect)
+    }
+
+    pub fn closest_point_to_origin(&self) -> Vec2 {
+        let max = self.max();
+        let mut min_dist = self.x.abs();
+        let mut bounds_point = Vec2::new(self.x, 0.0);
+
+        if max.x.abs() < min_dist {
+            min_dist = max.x.abs();
+            bounds_point.x = max.x;
+            bounds_point.y = 0.0;
+        }
+
+        if max.y.abs() < min_dist {
+            min_dist = max.y.abs();
+            bounds_point.x = 0.0;
+            bounds_point.y = max.y;
+        }
+
+        if self.y.abs() < min_dist {
+            // min_dist = self.y.abs();
+            bounds_point.x = 0.0;
+            bounds_point.y = self.y;
+        }
+
+        bounds_point
+    }
+
+    /// Returns (Closest, EdgeNormal)
+    pub fn closest_point_on_border(&self, point: Vec2) -> (Vec2, Vec2) {
+        let mut edge_normal = Vec2::ZERO;
+
+        let mut res = Vec2::new(0.0, 0.0);
+        res.x = point.x.clamp(self.left(), self.right());
+        res.y = point.y.clamp(self.top(), self.bottom());
+
+        if self.contains(res) {
+            let dl = res.x - self.left();
+            let dr = self.right() - res.x;
+            let dt = res.y - self.top();
+            let db = self.bottom() - res.y;
+
+            let min = dl.min(dr.min(dt.min(db)));
+            if min == dt {
+                res.y = self.top();
+                edge_normal.y = -1.0;
+            } else if min == db {
+                res.y = self.bottom();
+                edge_normal.y = 1.0;
+            } else if min == dl {
+                res.x = self.left();
+                edge_normal.x = -1.0;
+            } else {
+                res.x = self.right();
+                edge_normal.x = 1.0;
+            }
+        } else {
+            if res.x == self.left() {
+                edge_normal.x = -1.0;
+            }
+            if res.x == self.right() {
+                edge_normal.x = 1.0;
+            }
+            if res.y == self.top() {
+                edge_normal.y = -1.0;
+            }
+            if res.y == self.bottom() {
+                edge_normal.y = 1.0;
+            }
+        }
+
+        (res, edge_normal)
+    }
+
+    fn ray_intersects(&self, ray: &Ray2D) -> Option<f32> {
+        let mut distance = 0.0;
+        let mut max = f32::MAX;
+
+        if ray.direction.x.abs() < 1e-06 {
+            if ray.start.x < self.x || ray.start.x > self.x + self.width {
+                return None;
+            }
```

Comment:
```
if ray.direction.x.abs() < 1e-06 &&
             ray.start.x < self.x || ray.start.x > self.x + self.width {
             return None;
             }
```

?

Reply:
Your alternative is much harder to read imo